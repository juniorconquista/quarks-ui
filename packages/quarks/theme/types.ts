import {
  ColorProps,
  FlexboxProps,
  LayoutProps,
  SpaceProps,
  PositionProps,
  BorderProps,
  GridProps,
  FontSizeProps
} from 'styled-system'

export type CssProperties = {
  transition?: string
  filter?: string
  textTransform?:
  | 'none'
  | 'capitalize'
  | 'uppercase'
  | 'lowercase'
  | 'initial'
  | 'inherit'
  animationFillMode?:
  | 'none'
  | 'forwards'
  | 'backwards'
  | 'both'
  | 'initial'
  | 'inherit'
  objectFit?: 'contain' | 'cover' | 'fill' | 'none' | 'scale-down'
  whiteSpace?:
  | 'none'
  | 'nowrap'
  | 'pre'
  | 'pre-line'
  | 'pre-wrap'
  | 'initial'
  | 'inherit'
  '-webkit-box-shadow'?: string
  boxSizing?: string
}

export type DefaultStyleProps = BorderProps &
  ColorProps &
  FlexboxProps &
  FontSizeProps &
  LayoutProps &
  PositionProps &
  SpaceProps &
  GridProps & {
    // for some reason this isn't included; this is for replacing the 'base' component
    // as?: React.ComponentType;
  }
