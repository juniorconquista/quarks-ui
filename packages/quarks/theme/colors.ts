
// teste
export default (dark = false): any => ({
  dark,
  get primary () {
    return {
      default: this.dark ? '#375a7f' : '#007bff',
      contrast: this.dark ? '#2f75bf' : '#6aa5e4',
      dark: '#375a7f',
      light: '#007bff'
    }
  },
  get secondary () {
    return {
      default: this.dark ? '#444444' : '#6c757d',
      dark: '#444444',
      light: '#6c757d'
    }
  },
  get success () {
    return {
      default: this.dark ? '#00bc8c' : '#28a745',
      dark: '#00bc8c',
      light: '#28a745'
    }
  },
  get danger () {
    return {
      default: this.dark ? '#e74c3c' : '#dc3545',
      dark: '#e74c3c',
      light: '#dc3545'
    }
  },
  get warning () {
    return {
      default: this.dark ? '#f39c12' : '#ffc107',
      dark: '#f39c12',
      light: '#ffc107'
    }
  },
  get info () {
    return {
      default: this.dark ? '#3498db' : '#17a2b8',
      dark: '#3498db',
      light: '#17a2b8'
    }
  },
  get background () {
    return {
      default: this.dark ? '#f8f9fa' : '#343a40',
      dark: '#f8f9fa',
      light: '#343a40'
    }
  },
  get text () {
    return {
      default: this.dark ? '#f8f9fa' : '#343a40',
      dark: '#f8f9fa',
      light: '#343a40'
    }
  }
})
