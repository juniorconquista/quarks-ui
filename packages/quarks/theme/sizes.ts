export const breakpoints = ['0rem', '37.5rem', '62rem']

const spaceBase = 1 // 1rem
export const space = {
  none: '0',
  xs: `${spaceBase * 0.25}rem`,
  sm: `${spaceBase * 0.5}rem`,
  md: `${spaceBase}rem`,
  lg: `${spaceBase * 1.5}rem`,
  xl: `${spaceBase * 3}rem`
}

const fontSizeBase = 1 // 1rem
export const fontSizes = {
  base: `${fontSizeBase}rem`,
  sm: `${fontSizeBase * 0.875}rem`,
  lg: `${fontSizeBase * 1.5}rem`,
  h1: `${fontSizeBase * 2.5}rem`,
  h2: `${fontSizeBase * 2}rem`,
  h3: `${fontSizeBase * 1.75}rem`,
  h4: `${fontSizeBase * 1.5}rem`,
  h5: `${fontSizeBase * 1.25}rem`,
  h6: `${fontSizeBase}rem`
}

export const fontWeights = {
  thin: 100,
  light: 300,
  regular: 400,
  medium: 500,
  semiBold: 600,
  bold: 700,
  black: 900
}

export const radii = {
  none: '0',
  sm: '0.2rem',
  md: '0.25rem',
  lg: '0.3rem'
}
