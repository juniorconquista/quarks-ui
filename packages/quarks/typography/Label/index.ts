export { default } from './label'
export * from './label.types'
