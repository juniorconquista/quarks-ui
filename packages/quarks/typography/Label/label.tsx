import styled from 'styled-components'
import { variant, color, typography, space } from 'styled-system'
import { LabelProps } from './label.types'

const variants = variant({
  prop: 'styling',
  variants: {
    base: {
      color: 'text.default',
      fontSize: 4
    }
  }
})

export const Label: React.FC<LabelProps> = styled.label`
    ${variants}
    ${color}
    ${typography}
    ${space}
`

export default Label
