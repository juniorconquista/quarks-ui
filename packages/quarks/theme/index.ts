import colors from './colors'
import { breakpoints, fontSizes, fontWeights, radii, space } from './sizes'

const theme = {
  breakpoints,
  fontSizes,
  fontWeights,
  radii,
  space,
  colors: colors()
}

// @ts-expect-error
theme.breakpoints.mobile = breakpoints[0]
// @ts-expect-error
theme.breakpoints.tablet = breakpoints[1]
// @ts-expect-error
theme.breakpoints.desktop = breakpoints[2]

export default theme
