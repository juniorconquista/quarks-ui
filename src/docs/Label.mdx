---
name: Label
menu: Typography
---

import { Playground, Props } from 'docz';
import { Label } from '../ds/typography';

# Label

The _label_ element is the formal way of defining a label for an HTML form widget. This is the most important element if you want to create accessible forms - when implemented correctly, screen readers will speak the label of a form element along with related instructions.

### Best practices

## Properties

<Props of={Label} />

## Basic usage

<Playground>
    <Label styling="base">Base</Label>
</Playground>

## Customization

### Color

The color utility parses a component's _color_ and _bg_ props and converts them into CSS declarations. By default the raw value of the prop is returned. Color palettes can be configured with the ThemeProvider to use keys as prop values, with support for dot notation.

You can see all the color customization [here](https://styled-system.com/api/#color).

```javascript
// picks the value defined in `theme.colors.blue`
<Label styling="base" color='blue'>label</Label>

// picks up a nested color value using dot notation `theme.colors.gray[0]`
<Label color='gray.0'>label</Label>

// raw CSS color value
<Label color='#f00'>label</Label>

// background colors
<Label bg='blue'>label</Label>

// verbose prop
<Label backgroundColor='blue'>label</Label>
```

### Space

The space utility converts shorthand margin and padding props to _margin_ and _padding_ CSS declarations.

You can see all the space customization [here](https://styled-system.com/api#space).

```javascript
// sets margin value of `theme.space[2]`
<Label styling="base" m={2}>label</Label>

// sets a margin value of `16px` since it's greater than `theme.space.length`
<Label m={16}>label</Label>

// sets margin `'auto'`
<Label m='auto'>label</Label>

// sets margin `8px` fom mobile devices, `16px` for tablet and `32px` for desktop from the first breakpoint and up
<Label m={{ mobile:2, tablet:3, desktop:4 }}>label</Label>
```

### Typography

Typography props include _fontFamily_, _fontSize_, _fontWeight_, _lineHeight_, _letterSpacing_, _textAlign_, and _fontStyle_.

You can see all the typography customization [here](https://styled-system.com/api/#typography).

```javascript
// font-size of `theme.fontSizes[3]`
<Label styling="base" fontSize={3}>label</Label>

// font-size `32px`
<Label fontSize={32}>label</Label>

// font-size `'2em'`
<Label fontSize='2em'>label</Label>

// font-size `10px` on all viewports and `12px` from the first breakpoint and up
<Label fontSize={{ mobile:10, desktop:12 }}>label</Label>

// fontFamily
<Label fontFamily='mono'>label</Label>

// textAlign
<Label textAlign='center'>label</Label>
<Label textAlign={{ mobile: 'center', tablet: 'left', desktop: 'center' }}>label</Label>

// lineHeight
<Label lineHeight='1.25'>label</Label>

// fontWeight
<Label fontWeight='bold'>label</Label>

// letterSpacing
<Label letterSpacing='0.1em'>label</Label>
```
