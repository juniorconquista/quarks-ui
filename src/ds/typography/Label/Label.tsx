import styled from 'styled-components';
import { variant, VariantArgs, color, typography, space } from 'styled-system';
import { LabelStyledProps } from './label.types';

const variants = variant({
    prop: 'styling',
    variants: {
        base: {
            color: 'text',
            fontSize: 'fontSizes.4',
        },
    },
} as VariantArgs);

export const Label: LabelStyledProps = styled.label`
    ${variants}
    ${color}
    ${typography}
    ${space}
`;

export default Label;
