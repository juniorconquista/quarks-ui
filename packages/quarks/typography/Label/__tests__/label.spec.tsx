import React from 'react'
import renderer from 'react-test-renderer'
import ThemeTest from '../../../utils/ThemeProviderTest'
import 'jest-styled-components'

import Label from '../Label'
import theme from '../../../theme'

describe('Label', () => {
  test('basic snapshot', () => {
    const label = renderer
      .create(
                <ThemeTest>
                    <Label styling="base" />
                </ThemeTest>
      )
      .toJSON()
    expect(label).toMatchSnapshot()
  })

  test('should inherit the styling properties of the base variant', () => {
    const label = renderer
      .create(
                <ThemeTest>
                    <Label styling="base" />
                </ThemeTest>
      )
      .toJSON()
    expect(label).toHaveStyleRule('color', theme.colors.text.default)
    expect(label).toHaveStyleRule('font-size', theme.fontSizes[4])
  })

  test('should apply custom styling properties', () => {
    const label = renderer
      .create(<Label styling="base" mt="20px" p="3px" />)
      .toJSON()
    expect(label).toHaveStyleRule('margin-top', '20px')
    expect(label).toHaveStyleRule('padding', '3px')
  })
})
