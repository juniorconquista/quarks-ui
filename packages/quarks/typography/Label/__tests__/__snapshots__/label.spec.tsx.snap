// Jest Snapshot v1, https://goo.gl/fbAQLP

exports['Label basic snapshot 1'] = `
.c0 {
  color: #343a40;
  font-size: 1rem;
}

<label
  className="c0"
/>
`
